frontend/src/routes/Governance/discovered/grouping.ts

```diff
@@ -0,0 +1,205 @@
+    const getHostingSubscription = (annotations: any) =>
+      annotations['apps.open-cluster-management.io/hosting-subscription']
+    const parentAnnotations = annotations
+    let hostingSubscription = getHostingSubscription(parentAnnotations)
+    if (!hostingSubscription) {
+      // check if this policy was deployed by a Helm release
+      const releaseNamespace = parentAnnotations['meta.helm.sh/release-namespace']
+      const releaseName = parentAnnotations['meta.helm.sh/release-name']
+      if (releaseNamespace && releaseName) {
+        const helmReleaseMap = getResourceMap(helmReleases)
+        const helmRelease = helmReleaseMap[`${releaseNamespace}/${releaseName}`]
+        const helmReleaseAnnotations = getAnnotations(helmRelease)
+        hostingSubscription = getHostingSubscription(helmReleaseAnnotations)
+      }
+    }
+    if (hostingSubscription) {
+      const subscriptionMap = getResourceMap(subscriptions)
+      const channelMap = getResourceMap(channels)
+      const subscription = subscriptionMap[hostingSubscription]
+      const channel = channelMap[subscription?.spec.channel ?? '']
+      if (subscription && channel) {
+        return {
+          type: channel.spec.type ?? '',
+        }
+      }
+    }
+    return null
+  }
+
+  const getPolicySource = (policy: any, helmReleases: any[], channels: any[], subscriptions: any[]): ISourceType => {
+    const parentPolicy = getStringLabelValue(policy.label, 'policy.open-cluster-management.io/policy')
+    const parentNs = parentPolicy?.split('.')[0] ?? ''
+    const parentName = parentPolicy?.split('.')[1] ?? ''
+
+    if (parentNs && parentName) {
+      return {
+        type: 'Policy',
+        parentNs,
+        parentName,
+      }
+    }
+
+    const isExternal = policy._isExternal
+    let source: string | JSX.Element = 'Local'
+    if (isExternal) {
+      const policySource = resolveSource(policy.annotations ?? {}, helmReleases, channels, subscriptions)
+      source = policySource ? getSourceText(policySource, isExternal) : 'Managed externally'
+    }
+    // source is string
+    return {
+      type: source,
+      parentNs: '',
+      parentName: '',
+    }
+  }
+
+  enum PolicySeverity {
+    Unknown,
+    Low,
+    Medium,
+    High,
+    Critical,
+  }
+
+  const severityTable = (severity: string): number => {
+    if (!severity) return PolicySeverity.High
+
+    switch (severity.toLocaleLowerCase()) {
+      case 'low':
+        return PolicySeverity.Low
+      case 'medium':
+        return PolicySeverity.Medium
+      case 'high':
+        return PolicySeverity.High
+      case 'critical':
+        return PolicySeverity.Critical
+      default:
+        // unknown or null
+        return PolicySeverity.High
+    }
+  }
+
+  const createMessage = (data: any[], helmReleases: any[], channels: any[], subscriptions: any[]): any[] => {
+    const policiesWithSource = JSON.parse(JSON.stringify(data), (_, v: string) => {
+      if (v === 'true') return true
```

Comment:
I'm not sure this is specific enough. Could you please have an allowlist of keys that you are transforming?

Reply:
@yiraeChristineKim right, but I think this function gets called on nested objects too. Can you change the code to only transform the boolean fields you need to not encounter anything unexpected?